src/tiled/objectselectiontool.cpp

```diff
@@ -551,8 +552,22 @@ void ObjectSelectionTool::mouseReleased(QGraphicsSceneMouseEvent *event)
             break;
         }
         const Qt::KeyboardModifiers modifiers = event->modifiers();
+        QSet<MapObjectItem*> selection = mapScene()->selectedObjectItems();
+        if (modifiers & Qt::AltModifier) {
+            auto underlyingObjects = listOfObjectItemsAt(event->scenePos());
+
+            if (!underlyingObjects.size())
+                break;
+            if (mSelectedRotationIndex >= underlyingObjects.size())
+                mSelectedRotationIndex = 0;//definitely another stack
```

Comment:
I think instead of keeping track of the current "selection rotation index", the next selection should simply be based on the current selection. So basically, find the last selected object among the objects at the mouse position, and then take the next one (or cycle back to the first one if there is no next one).

It will be a bit more processing but I think it will make the system more predictable. At the moment there are some problems with the click sometimes not responding the first time you try it, or a deeper object getting selected than you might expect after having already used the feature somewhere else.